CSnake uses wrong third party folder on Windows with inconsistent letter casing in paths.

The test execution on Jenkins is broken on the Windows 10 slave. The test "_csnBuildDummyLibTests.csnBuildDummyLibTests.testDummyLibBuildMultiple_" fails because of the following CMake error:

```
CMake Error at C:/Users/Jenkins/workspace/CSnake-win10/tests/build/executable/DummyLib2Applications_myApp/CMakeLists.txt:24 (INCLUDE):
  include could not find load file:

    C:/Users/Jenkins/workspace/CSnake-win10/tests/build/thirdParty/Five/FiveConfig.cmake
```

--- a/src/csnProject.py
+++ b/src/csnProject.py
@@ -188,14 +188,16 @@ class GenericProject(VeryGenericProject):
             _sourceRootFolder = csnUtility.NormalizePath(os.path.dirname(FindFilename(1)))
         VeryGenericProject.__init__(self, _name, _type, _sourceRootFolder, _categories, _context)
         
-        # TODO: Remove this code in CSnake 3.0. Then this process is done exclusively in the subclass for third parties.
-        self.thirdPartyIndex = 0
+        # TODO: Remove this code as soon as possible. This process should be done exclusively in the subclass for third parties.
+        self.thirdPartyIndex = -1
         count = 0
         for folder in self.context.GetThirdPartyFolders():
-            if csnUtility.NormalizePath(folder) == os.path.dirname(_sourceRootFolder):
+            if csnUtility.IsSameFileOrDirectory(folder, os.path.dirname(_sourceRootFolder)):
                 self.thirdPartyIndex = count
                 break
             count += 1
+        if self.thirdPartyIndex < 0 and self.type == "third party":
+            raise Exception("Could not find any thirdparty folder for this project.")
         
         self.rules = dict()
         self.customCommands = []
@@ -443,13 +445,15 @@ class ThirdPartyProject(VeryGenericProject):
         VeryGenericProject.__init__(self, name, "third party", sourceRootFolder, None, context)
         
         # Get the thirdPartyBuildFolder index
-        self.thirdPartyIndex = 0
+        self.thirdPartyIndex = -1
         count = 0
         for folder in self.context.GetThirdPartyFolders():
-            if csnUtility.NormalizePath(folder) == os.path.dirname(sourceRootFolder):
+            if csnUtility.IsSameFileOrDirectory(folder, os.path.dirname(sourceRootFolder)):
                 self.thirdPartyIndex = count
                 break
             count += 1
+        if self.thirdPartyIndex < 0:
+            raise Exception("Could not find any thirdparty folder for this project.")
     
     def GetBuildFolder(self):
         return self.context.GetThirdPartyBuildFolderByIndex(self.thirdPartyIndex)

--- a/src/csnUtility.py
+++ b/src/csnUtility.py
@@ -5,7 +5,6 @@ import sys
 import GlobDirectoryWalker
 import shutil
 import inspect
-import os
 import os.path
 import logging.config
 import string
@@ -34,10 +33,19 @@ def IsSameFileOrDirectory(path1, path2):
     Returns True, if both paths are referring to the same physical file or directory.
     Returns False, if they refer to different files/directories or at least one file or directory doesn't exist.
     """
-    try:
-        return os.stat(path1) == os.stat(path2)
-    except:
-        return False
+    if hasattr(os.path, "samefile"):
+        # check based on "os.path.samefile", which checks the file system inode
+        # this is pretty robust (against multiple mounts, symlinks, etc.),
+        # as long as the file system is local (we can't get a remote server's
+        # internal inode, so this may e.g. report false negatives on SMB mounts)
+        try:
+            return os.path.samefile(path1, path2)
+        except:
+            return False
+    else:
+        # The best we can do without "os.path.samefile":
+        # Normalize the format and letter case of the string and compare the results
+        return NormalizePath(path1, True) == NormalizePath(path2, True)
 
 def NormalizePath(path, _correctCase = True):
     path = os.path.normpath(path)

--- a/src/csnUtility.py
+++ b/src/csnUtility.py
@@ -13,11 +13,20 @@ if sys.platform == 'win32':
 if sys.platform != 'win32':
     import commands
 
-def CorrectPath(path):
+
+def CorrectPathCase(path):
+    # Note: It might (?) be more robust/efficient to use "os.path.normcase"
+    # here, but that might break backwards compatibility.
+    # The function "os.path.normcase" normalizes always to lower case on
+    # case sensitive systems, whereas "CorrectPathCase" normalizes to the
+    # "real" letter casing in the file system. Old config files may still
+    # contain the old normalized values, which may break the comparison
+    # with new normalized values.
+    
     (first,second) = os.path.split(path)
     
     if second != "":
-        firstCorrected = CorrectPath(first)
+        firstCorrected = CorrectPathCase(first)
         secondCorrected = second
         if os.path.exists(first):
             for name in os.listdir(first):
@@ -50,7 +59,7 @@ def IsSameFileOrDirectory(path1, path2):
 def NormalizePath(path, _correctCase = True):
     path = os.path.normpath(path)
     if _correctCase:
-        path = CorrectPath(path)
+        path = CorrectPathCase(path)
     path = path.replace("\\", "/")
     return path
 

--- a/src/csnUtility.py
+++ b/src/csnUtility.py
@@ -5,6 +5,7 @@ import sys
 import GlobDirectoryWalker
 import shutil
 import inspect
+import os
 import os.path
 import logging.config
 import string
@@ -26,6 +27,17 @@ def CorrectPath(path):
         return os.path.join(firstCorrected, secondCorrected)
     else:
         return first
+
+def IsSameFileOrDirectory(path1, path2):
+    """
+    Checks, if two paths are referring to the same physical file or directory.
+    Returns True, if both paths are referring to the same physical file or directory.
+    Returns False, if they refer to different files/directories or at least one file or directory doesn't exist.
+    """
+    try:
+        return os.stat(path1) == os.stat(path2)
+    except:
+        return False
 
 def NormalizePath(path, _correctCase = True):
     path = os.path.normpath(path)

--- a/tests/csnUtilityTests.py
+++ b/tests/csnUtilityTests.py
@@ -32,27 +32,27 @@ class csnUtilityTests(unittest.TestCase):
         self.assertTrue( csnUtility.HasBackSlash("c:\\hallo") )
         self.assertFalse( csnUtility.HasBackSlash("c://hallo") )
         
-    def testCorrectPath(self):
+    def testCorrectPathCase(self):
         """ csnUtilityTests: test CorrectPath function. """
         # check where we are
         root = "data/my src/"
         refPathRoot = root + "DummyLib"
         refPath = os.path.normpath( refPathRoot + "/libmodules" )
         
-        self.assertEqual( csnUtility.CorrectPath(""), "" )
-        self.assertEqual( csnUtility.CorrectPath("."), "." )
+        self.assertEqual( csnUtility.CorrectPathCase(""), "" )
+        self.assertEqual( csnUtility.CorrectPathCase("."), "." )
         testPath1 = root + "DummyLib/libmodules"
-        self.assertEqual( csnUtility.CorrectPath(testPath1), refPath )
+        self.assertEqual( csnUtility.CorrectPathCase(testPath1), refPath )
         testPath2 = root + "DummyLib/liBmoDules"
-        self.assertEqual( csnUtility.CorrectPath(testPath2), refPath )      
+        self.assertEqual( csnUtility.CorrectPathCase(testPath2), refPath )      
         testPath3 = root + "DuMMyLib/libmodules"
-        self.assertEqual( csnUtility.CorrectPath(testPath3), refPath )      
+        self.assertEqual( csnUtility.CorrectPathCase(testPath3), refPath )      
         refPath4 = os.path.normpath( root + "DummyLib/doEsnoTexist" )
         testPath4 = root + "DuMMyLib/doEsnoTexist"
-        self.assertEqual( csnUtility.CorrectPath(testPath4), refPath4 )        
+        self.assertEqual( csnUtility.CorrectPathCase(testPath4), refPath4 )        
         refPath5 = os.path.normpath( "doEs/nOt/eXist" )
         testPath5 = "doEs/nOt/eXist"
-        self.assertEqual( csnUtility.CorrectPath(testPath5), refPath5 )  
+        self.assertEqual( csnUtility.CorrectPathCase(testPath5), refPath5 )  
         
     def testSearchProgramPath(self):
         if csnUtility.IsWindowsPlatform():

